battlemogs-sage/src/transitions/mod.rs

```diff
@@ -0,0 +1,237 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{error::*, BattleMogsTransition};
+
+use ajuna_payment_handler::NativeId;
+use ajuna_primitives::sage_api::SageApi;
+use sage_api::{traits::TransitionOutput, TransitionError};
+
+use crate::asset::{BattleMogsAsset, BattleMogsId};
+use frame_support::{
+	ensure,
+	pallet_prelude::{Decode, Encode, TypeInfo},
+	Parameter,
+};
+use parity_scale_codec::{Codec, MaxEncodedLen};
+use sp_core::H256;
+use sp_runtime::{
+	traits::{AtLeast32BitUnsigned, BlockNumber as BlockNumberT, Member},
+	SaturatedConversion,
+};
+
+mod breed;
+mod create;
+mod hatch;
+mod morph;
+mod register;
+mod remove;
+mod sacrifice;
+mod sarifice_into;
+
+pub(crate) type BattleMogsTransitionOutput<BlockNumber> =
+	Vec<TransitionOutput<BattleMogsId, BattleMogsAsset<BlockNumber>>>;
+
+#[derive(Encode, Decode, Debug, Copy, Clone, PartialEq, Eq, TypeInfo)]
+pub enum BreedType {
+	DomDom = 0,
+	DomRez = 1,
+	RezDom = 2,
+	RezRez = 3,
+}
+
+impl BreedType {
+	fn calculate_breed_type<BlockNumber>(block_number: BlockNumber) -> BreedType
+	where
+		BlockNumber: BlockNumberT,
+	{
+		let mod_value: u32 = 80;
+		let modulo_80 = (block_number % mod_value.into()).saturated_into::<u32>();
+
+		match modulo_80 {
+			0..=19 => BreedType::DomDom,
+			20..=39 => BreedType::DomRez,
+			40..=59 => BreedType::RezDom,
+			_ => BreedType::RezRez,
+		}
+	}
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Encode, Decode, MaxEncodedLen, TypeInfo)]
+pub struct BattleMogsTransitionConfig {
+	pub max_mogwais: u16,
+	pub target_egg_hatcher: u16,
+	pub target_sacrificer: u16,
+	pub target_morpheus: u16,
+	pub target_legend_breeder: u16,
+	pub target_promiscuous: u16,
+}
+
+pub const DEFAULT_MAX_MOGWAIS: u16 = 10;
+pub const DEFAULT_TARGET: u16 = 100;
+
+impl Default for BattleMogsTransitionConfig {
+	fn default() -> Self {
+		Self {
+			max_mogwais: DEFAULT_MAX_MOGWAIS,
+			target_egg_hatcher: DEFAULT_TARGET,
+			target_sacrificer: DEFAULT_TARGET,
+			target_morpheus: DEFAULT_TARGET,
+			target_legend_breeder: DEFAULT_TARGET,
+			target_promiscuous: DEFAULT_TARGET,
+		}
+	}
+}
+
+impl<AccountId, BlockNumber, Balance, Sage> BattleMogsTransition<AccountId, BlockNumber, Sage>
+where
+	AccountId: Member + Codec,
+	BlockNumber: BlockNumberT,
+	Balance: Member + Parameter + AtLeast32BitUnsigned + MaxEncodedLen,
+	Sage: SageApi<
+		AccountId = AccountId,
+		AssetId = BattleMogsId,
+		Asset = BattleMogsAsset<BlockNumber>,
+		Balance = Balance,
+		BlockNumber = BlockNumber,
+		TransitionConfig = BattleMogsTransitionConfig,
+		HashOutput = H256,
+	>,
+{
+	pub(crate) fn new_asset_id(subject: &[u8], nonce: u64) -> BattleMogsId {
+		Sage::random_hash(subject).to_low_u64_be().saturating_add(nonce)
```

Comment:
This is where we need to be careful, where does the `nonce` come from?

Shouldn't we at least check that the asset id does not exist yet?

Reply:
Fundamentally we have to trust the `Sage::random_hash` method in here

Reply:
Hmm, but as we said yesterday; The random hash is the same for the entirety of a block, and we either:

* Declare that very well in the api, and users of it need to be aware that for different outputs, they need to add different subjects.
* Add some state to the randomness to have an internally changing subject that is appended to the user's input subject. This variant, however, can be manipulated by transaction reordering (not that bad as we are the only running collators). Still, the more stupid our randomness is, the better.